### PR TITLE
Detect and register event bus codec automatically

### LINKED ISSUE #2529: Detect and register event bus codec automatically
**Description**
Detect and register event bus codec automatically

**Implementation ideas**
We need to collect `@ConsumeEvent` method and check for received type:
* direct payload (so type T)
* Message<T>

Once the types are collected, a `GenericCodec` must be registered. It the `@ConsumeEvent` specified a specific behavior this one must be registered.  

The generic codec does the following:

* pass the received instance
* use `clone()` if available

At the same time, the `local` attribute of `ConsumeEvent` should be passed to `true` by default.

### PR DESCRIPTION
**Pull request by Dufgui**
fix #2529

### CODE REVIEW

**Comment by aloubyansky** [comment]
The build failed due to source code formatting rules violations.

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/GenericMessageCodec.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.lang.reflect.Method;
+
+import org.jboss.logging.Logger;
+
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.eventbus.MessageCodec;
+import io.vertx.core.json.JsonObject;
+
+public class GenericMessageCodec implements MessageCodec<JsonObject, JsonObject> {
```
Comment (cescoffier) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/GenericMessageCodec.java]: This would be pretty expensive. In the case of local delivery, don't serialize to Json.

The codec should be `Object,Object`.

File: extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/VertxProcessor.java [hunk 2]
```diff
@@ -329,4 +342,69 @@ private void invoke(BeanInfo bean, MethodInfo method, ResultHandle messageHandle
...
+        } catch (IllegalAccessException e) {
+            LOGGER.warn("Cannot instantiate the MessageCodec " + codec.toString() + " because " + e.getMessage());
+        }
+
+    }
+
+    private void registerCodec(Type typeToAdd, MessageCodec codec) {
+        final Vertx vertx = Vertx.vertx();
+        EventBus eventBus = vertx.eventBus();
+        eventBus.registerDefaultCodec(typeToAdd.name().getClass(), codec);
```
Comment (cescoffier) [inline extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/VertxProcessor.java]: you are not registering the codec on the right event bus. They need to be registered on the Vert.x instance that is injected.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/GenericMessageCodec.java (hunk 1 above)
Comment (Dufgui) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/GenericMessageCodec.java]: Ok but I have a buffer, how do you want to serialize them on it ? 
If I don't use Json, I can serializable (ObjectOutputStream), or any other kind of conversion to byte array. because the signature use a buffer.  no ?

Comment (Dufgui) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/GenericMessageCodec.java]: i push a version based on serialization. If you are ok, i will add unit test after the implementation is ok for you

Comment (Dufgui) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/GenericMessageCodec.java]: Is it ok for you @cescoffier ?

**Comment by Dufgui** [comment]
Hello, I need help on this, sorry.
I have an error during my test on missing class CompletionStage. It seems to be in the recorder. But I don't really understand why.

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.223 s <<< FAILURE! - in io.quarkus.vertx.MessageConsumerFailureTest
[ERROR] io.quarkus.vertx.MessageConsumerFailureTest  Time elapsed: 0.223 s  <<< ERROR!
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:279)
        at ...
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl3.doStart(ApplicationImpl3.zig:196)
        at io.quarkus.runtime.Application.start(Application.java:84)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:127)
        at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:244)
        ... 35 more
Caused by: java.lang.ClassNotFoundException: java.util.concurrent.CompletionStage<java.lang.String>
        at io.quarkus.runner.RuntimeClassLoader.findClass(RuntimeClassLoader.java:229)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at io.quarkus.runner.RuntimeClassLoader.loadClass(RuntimeClassLoader.java:166)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:348)
        at io.quarkus.deployment.steps.VertxProcessor$build26.deploy_0(VertxProcessor$build26.zig:169)
        at io.quarkus.deployment.steps.VertxProcessor$build26.deploy(VertxProcessor$build26.zig:36)
        at io.quarkus.runner.ApplicationImpl3.doStart(ApplicationImpl3.zig:168)
        ... 38 more
```
Could you explain my why if you understand to fix it please?

**Comment by Dufgui** [comment]
So you can review ;o) @cescoffier

File: extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/MessageCodecItem.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package io.quarkus.vertx.deployment;
+
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class MessageCodecItem extends MultiBuildItem {
```
Comment (gsmet) [inline extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/MessageCodecItem.java]: This class should be called `MessageCodecBuildItem`.

File: extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/VertxProcessor.java [hunk 4]
```diff
@@ -127,7 +137,8 @@ VertxBuildItem build(VertxRecorder recorder, BeanContainerBuildItem beanContaine
             List<EventConsumerBusinessMethodItem> messageConsumerBusinessMethods,
             BuildProducer<GeneratedClassBuildItem> generatedClass,
             AnnotationProxyBuildItem annotationProxy, LaunchModeBuildItem launchMode, ShutdownContextBuildItem shutdown,
-            VertxConfiguration config, BuildProducer<ServiceStartBuildItem> serviceStart) {
+            VertxConfiguration config, BuildProducer<ServiceStartBuildItem> serviceStart,
+            List<MessageCodecItem> messageCodecItems, RecorderContext recorderContext) {
```
Comment (gsmet) [inline extensions/vertx/deployment/src/main/java/io/quarkus/vertx/deployment/VertxProcessor.java]: for consistency, the variable should be called `messageCodecs`

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/ConsumeEvent.java [hunk 5]
```diff
@@ -63,7 +66,7 @@
      * @return {@code true} if the address should not be propagated across the cluster
      * @see io.vertx.core.eventbus.EventBus#localConsumer(String)
      */
-    boolean local() default false;
+    boolean local() default true;
```
Comment (gsmet) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/ConsumeEvent.java]: Hmmm, you change the default value here. Is it something we want? If so, it should at least be a separate commit if not a separate PR.

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxRecorder.java [hunk 6]
```diff
@@ -319,4 +323,28 @@ private EventConsumerInvoker createInvoker(String invokerClassName) {
...
+    private void registerCodec(Class<?> typeToAdd, Class<?> messageCodecClass) {
+        try {
+            if (messageCodecClass.isAssignableFrom(MessageCodec.class)) {
+                MessageCodec messageCodec = (MessageCodec) messageCodecClass.newInstance();
+                registerCodec(typeToAdd, messageCodec);
+            } else {
+                LOGGER.warn(String.format("The codec %s does not inherit from MessageCodec ", messageCodecClass.toString()));
+            }
+        } catch (InstantiationException | IllegalAccessException e) {
+            LOGGER.warn("Cannot instantiate the MessageCodec " + messageCodecClass.toString() + " because " + e.getMessage());
```
Comment (gsmet) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxRecorder.java]: Should it be a warn or should we throw an error? In any case, I think the stacktrace shouldn't be lost.

**Review by gsmet — DISMISSED** [review]
I have made a quick review of the code (I'm no Clément so he will have to check if it's what we want, I just reviewed the code itself).

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/ConsumeEvent.java (hunk 5 above)
Comment (Dufgui) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/ConsumeEvent.java]: it's explicity ask in  the ticket: #2529

"At the same time, the local attribute of ConsumeEvent should be passed to true by default."

**Comment by Dufgui** [comment]
maybe the idea is to use this PR : https://github.com/quarkusio/quarkus/pull/3553 ?

**Comment by gsmet** [comment]
@cescoffier I let you provide feedback on this one or approve it?

**Review by cescoffier — APPROVED** [review]
